### PR TITLE
Performance improvements for the Delta Lake history system table 

### PR DESCRIPTION
**Pull request by alexjo2144**
## Description

The $history table does not need to lead the entire transaction log when a version filter is supplied.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Improve performance of the $history system table.
```

### CODE REVIEW

**Comment by findinpath** [comment]
FYI I have a PR opened a few months ago on a related topic  - https://github.com/trinodb/trino/pull/16192

**Comment by alexjo2144** [comment]
> FYI I have a PR opened a few months ago on a related topic - https://github.com/trinodb/trino/pull/16192

Which I did not review.... sorry for missing it. Should I close this?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java [hunk 1]
```diff
@@ -155,6 +155,55 @@ public void testReadWholePartition()
         assertUpdate("DROP TABLE test_read_part_key");
     }
 
+    @Test
+    public void testHistorySystemTable()
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java:159]: Please add to test `io.trino.plugin.deltalake.TestDeltaLakeSystemTables` also a test verifying the accuracy of the filters on `$history` table.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java [hunk 2]
```diff
@@ -155,6 +155,55 @@ public void testReadWholePartition()
...
+        assertFileSystemAccesses("SELECT * FROM \"test_history_system_table$history\" WHERE version >= 3 OR version = 1",
+                ImmutableMultiset.<FileOperation>builder()
+                        .add(new FileOperation(TRANSACTION_LOG_JSON, "00000000000000000001.json", INPUT_FILE_NEW_STREAM))
+                        .add(new FileOperation(TRANSACTION_LOG_JSON, "00000000000000000002.json", INPUT_FILE_NEW_STREAM))
+                        .add(new FileOperation(TRANSACTION_LOG_JSON, "00000000000000000003.json", INPUT_FILE_NEW_STREAM))
+                        .add(new FileOperation(TRANSACTION_LOG_JSON, "00000000000000000004.json", INPUT_FILE_NEW_STREAM))
+                        .add(new FileOperation(TRANSACTION_LOG_JSON, "00000000000000000005.json", INPUT_FILE_NEW_STREAM))
+                        .build());
+
+        assertFileSystemAccesses("SELECT * FROM \"test_history_system_table$history\" WHERE version >= 1 AND version < 3",
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java:197]: I commented all the  ..$history  `assertFileSystemAccesses` checks and got the following error:

```
java.lang.AssertionError: Expected: 
		24 fewer occurrences of FileOperation[fileType=LAST_CHECKPOINT, fileId=_last_checkpoint, operationType=INPUT_FILE_NEW_STREAM],
		24 fewer occurrences of FileOperation[fileType=TRANSACTION_LOG_JSON, fileId=00000000000000000005.json, operationType=INPUT_FILE_NEW_STREAM]
```

Comment (findinpath) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java:197]: See also on CI that `trino-delta-lake` is actually failing.

**Comment by findinpath** [comment]
> > FYI I have a PR opened a few months ago on a related topic - https://github.com/trinodb/trino/pull/16192

> Which I did not review.... sorry for missing it. Should I close this?

My PR is rather outdated. 

I'm thinking that the idea of an iterator over the transaction log entries is worth pursuing.

https://github.com/trinodb/trino/pull/16192/files#diff-9d037e6fdb6afb57337f6259ef3fc8c33f5b1c8e23b491070306ab1e9c08e902

**Comment by alexjo2144** [comment]
Pinned this issue down today, looks like `getSystemTable` gets called a whole bunch of times for a single query so any file system operations we do there get done ~22 times. For now I'm moving most of them to be lazy but they shouldn't need to be.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeHistoryTable.java [hunk 3]
```diff
@@ -85,13 +96,54 @@ public ConnectorTableMetadata getTableMetadata()
...
     {
-        if (commitInfoEntries.isEmpty()) {
-            return new FixedPageSource(ImmutableList.of());
+        Optional<Long> startVersion = Optional.empty();
+        Optional<Long> endVersion = Optional.empty();
+
+        if (constraint.getDomains().isPresent()) {
+            Map<Integer, Domain> domains = constraint.getDomains().get();
+            if (domains.containsKey(0)) {
+                Domain versionDomain = domains.get(0); // The zero value here relies on the column ordering defined in the constructor
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeHistoryTable.java:105]: You could add
```
versionColumnIndex = IntStream.range(0, tableMetadata.getColumns().size())
                .filter(i -> tableMetadata.getColumns().get(i).getName().equals("version"))
                .boxed()
                .collect(onlyElement());
```
in the constructor and remove the code comment

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 4]
```diff
@@ -3138,28 +3138,32 @@ private Optional<SystemTable> getRawSystemTable(ConnectorSession session, Schema
...
-        catch (TrinoException e) {
-            if (e.getErrorCode().equals(DELTA_LAKE_INVALID_SCHEMA.toErrorCode())) {
-                return Optional.empty();
-            }
-            throw e;
-        }
-
         return switch (tableType.get()) {
             case DATA -> throw new VerifyException("Unexpected DATA table type"); // Handled above.
-            case HISTORY -> Optional.of(new DeltaLakeHistoryTable(
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:3155]: nit: unnecessary formatting change?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 5]
```diff
@@ -3138,28 +3138,32 @@ private Optional<SystemTable> getRawSystemTable(ConnectorSession session, Schema
         }
 
         String tableLocation = table.get().location();
-        TableSnapshot tableSnapshot = getSnapshot(new SchemaTableName(systemTableName.getSchemaName(), tableName), tableLocation, session);
-        MetadataEntry metadataEntry;
-        try {
-            metadataEntry = transactionLogAccess.getMetadataEntry(tableSnapshot, session);
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:3144]: Moving this into `case PROPERTIES` could be a separate commit. It would be easier to understand that change isn't related to other changes (it took me a while to understand that)

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeHistoryTable.java [hunk 6]
```diff
@@ -85,13 +96,54 @@ public ConnectorTableMetadata getTableMetadata()
     @Override
     public ConnectorPageSource pageSource(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
     {
-        if (commitInfoEntries.isEmpty()) {
-            return new FixedPageSource(ImmutableList.of());
+        Optional<Long> startVersion = Optional.empty();
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeHistoryTable.java:99]: startVersion has natural default value (0), so it doesn't need to be optional

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeHistoryTable.java [hunk 7]
```diff
@@ -85,13 +96,54 @@ public ConnectorTableMetadata getTableMetadata()
...
+                }
+                else {
+                    Optional<Long> lowValue = range.getLowValue().map(Long.class::cast);
+                    if (lowValue.isPresent()) {
+                        startVersion = Optional.of(lowValue.get() - (range.isLowInclusive() ? 1 : 0));
+                    }
+
+                    Optional<Long> highValue = range.getHighValue().map(Long.class::cast);
+                    if (highValue.isPresent()) {
+                        endVersion = Optional.of(highValue.get() - (range.isHighInclusive() ? 0 : 1));
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeHistoryTable.java:120]: rename startVersion and endVersion to be either `*Inclusive` or `Exclusive` (startVersionExclusive and endVersionInclusive, right?)

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java [hunk 8]
```diff
@@ -542,7 +542,7 @@ private void testCorruptedTableLocation(String tableName, Path tableLocation, bo
 
         // Assert queries fail cleanly
         assertQueryFails("TABLE " + tableName, "Metadata not found in transaction log for tpch." + tableName);
-        assertQueryFails("SELECT * FROM \"" + tableName + "$history\"", ".* Table '.*\\$history' does not exist");
+        assertThat(query("SELECT * FROM \"" + tableName + "$history\"")).returnsEmptyResult();
```
Comment (findepi) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:545]: Why it no longer fails?

**Review by findepi — APPROVED** [review]

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java (hunk 8 above)
Comment (alexjo2144) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:545]: This is a side effect of not fetching the metadata in `DeltaLakeMetadata#getSystemTable` for history tables.

The expectations of what to do for corrupted tables is not super clear to me but empty results instead of an error for querying system tables didn't seem like a problematic behavior change to me.

Comment (findepi) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeBasic.java:545]: does it mean we produce no rows even if transaction log not found at all?
sounds like too much short-circuiting. i guess easy to address
